Source/TrackableEntities.EF.5/DbContextExtensions.cs

```diff
@@ -624,12 +657,82 @@ private static EntityState ToEntityState(this TrackingState state)
             }
         }
 
-        enum RelationshipType
+        private static RelationshipType GetRelationshipType(this DbContext dbContext, Type entityType, string propertyName)
         {
-            ManyToOne,
-            OneToOne,
-            ManyToMany,
-            OneToMany
+            // Get navigation property
+            var edmEntityType = dbContext.GetEdmSpaceType(entityType);
+            if (edmEntityType == null)
+                throw new ArgumentException("Getting entity type from metadata failed.", "entityType");
+            var navProp = edmEntityType.NavigationProperties
+                .SingleOrDefault(p => p.Name == propertyName);
+            if (navProp == null)
+                throw new ArgumentException("Getting navigation property failed.", "propertyName");
+
+            if (navProp.FromEndMember.RelationshipMultiplicity == RelationshipMultiplicity.One
+                && (navProp.ToEndMember.RelationshipMultiplicity == RelationshipMultiplicity.ZeroOrOne
+                    || navProp.ToEndMember.RelationshipMultiplicity == RelationshipMultiplicity.One))
+                return RelationshipType.OneToOne;
+
+            if (navProp.FromEndMember.RelationshipMultiplicity == RelationshipMultiplicity.Many
+                && (navProp.ToEndMember.RelationshipMultiplicity == RelationshipMultiplicity.ZeroOrOne
+                    || navProp.ToEndMember.RelationshipMultiplicity == RelationshipMultiplicity.One))
+                return RelationshipType.ManyToOne;
+
+            if (navProp.FromEndMember.RelationshipMultiplicity == RelationshipMultiplicity.Many
+                && navProp.ToEndMember.RelationshipMultiplicity == RelationshipMultiplicity.Many)
+                return RelationshipType.ManyToMany;
+
+            if ((navProp.FromEndMember.RelationshipMultiplicity == RelationshipMultiplicity.One
+                 || navProp.FromEndMember.RelationshipMultiplicity == RelationshipMultiplicity.ZeroOrOne)
+                && navProp.ToEndMember.RelationshipMultiplicity == RelationshipMultiplicity.Many)
+                return RelationshipType.OneToMany;
+
+            throw new InvalidOperationException(String.Format("Cannot determine relationship type for {0} property on {1}.", propertyName, entityType.FullName));
```

Comment:
R# will be happy, but I will be sad :wink: Just kidding. When I call static methods of `String` object, I preffer to use `String` instead of `string`, becouse it really is a class. Opposite to `string str = ""` is better than `String str = ""`. But ok, I'll change it. BTW I am preparing full R# defaults. I'll write about it later and not in this PR...